pennylane/devices/qubit/adjoint_jacobian.py

```diff
@@ -0,0 +1,130 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Functions to apply adjoint jacobian differentiation"""
+
+import numpy as np
+
+import pennylane as qml
+
+from pennylane.operation import operation_derivative
+from pennylane.tape import QuantumTape
+
+from .apply_operation import apply_operation
+
+from .initialize_state import create_initial_state
+
+# pylint: disable=protected-access, too-many-branches
+
+
+def adjoint_jacobian(
+    tape: QuantumTape,
+    prep_operation: qml.operation.StatePrep = None,
+    # starting_state=None,
+    # ToDo: is it okay that this is removed?
+):  # pylint: disable=too-many-statements
+    """Implements the adjoint method outlined in
+    `Jones and Gacon <https://arxiv.org/abs/2009.02823>`__ to differentiate an input tape.
+
+    After a forward pass, the circuit is reversed by iteratively applying adjoint
+    gates to scan backwards through the circuit.
+
+    .. note::
+        The adjoint differentiation method has the following restrictions:
+
+        * As it requires knowledge of the statevector, only statevector simulator devices can be
+          used.
+
+        * Only expectation values are supported as measurements.
+
+        * Does not work for parametrized observables like
+          :class:`~.Hamiltonian` or :class:`~.Hermitian`.
+
+    Args:
+        tape (.QuantumTape): circuit that the function takes the gradient of
+
+    Keyword Args:
+        starting_state (tensor_like): post-forward pass state to start execution with. It should be
+            complex-valued. Takes precedence over ``use_device_state``.
+        use_device_state (bool): use current device state to initialize. A forward pass of the same
+            circuit should be the last thing the device has executed. If a ``starting_state`` is
+            provided, that takes precedence.
+
+    Returns:
+        array or tuple[array]: the derivative of the tape with respect to trainable parameters.
+        Dimensions are ``(len(observables), len(trainable_params))``.
+
+    Raises:
+        QuantumFunctionError: if the input tape has measurements that are not expectation values
+            or contains a multi-parameter operation aside from :class:`~.Rot`
+    """
+    # pylint: disable=unnecessary-lambda-assignment
+
+    # Initialization of state
+    ket = create_initial_state(
+        wires=tape.wires, prep_operation=prep_operation
+    )  #  ket(0) if prep_operation is None, else
+    for op in tape.operations:
+        ket = apply_operation(op, ket)
+
+    # ToDo: compare to demo, make sure it all makes sense
+
+    # currently assumes only one observable, no batching
+    bra = apply_operation(tape.observables[0], ket)
+
+    expanded_ops = []
+    for op in reversed(tape.operations):
+        if op.num_params > 1:
+            ops = op.decomposition()
+            expanded_ops.extend(reversed(ops))  # ToDo: why do these need to be expanded?
```

Comment:
This will now be done in `DefaultQubit2.preprocess`. We do not have to handle this here anymore.